src/check_presence.c

```diff
@@ -0,0 +1,533 @@
+  pthread_cond_broadcast(&presence_cond);
+  pthread_mutex_unlock(&presence_lock);
+
+  int status = pthread_join(presence_thread_id, /* return = */ NULL);
+  if (status != 0) {
+    ERROR(PLUGIN_NAME " plugin: stopping thread failed.");
+    status = -1;
+  }
+
+  pthread_mutex_lock(&presence_lock);
+  memset(&presence_thread_id, 0, sizeof(presence_thread_id));
+  pthread_mutex_unlock(&presence_lock);
+
+  return status;
+} /* }}} int stop_thread */
+
+static int presence_init(void) {
+
+  if (!state_datastore) {
+    return 0;
+  }
+
+  int rc = mdb_env_create(&env);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_env_create failed: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  rc = mdb_env_open(env, state_datastore, MDB_NOSUBDIR, 0664);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: opening path '%s' failed: %s (%d)",
+          state_datastore, mdb_strerror(rc), rc);
+    mdb_env_close(env);
+    env = NULL;
+    return -1;
+  }
+
+  return start_thread();
+}
+
+static int presence_shutdown(void) {
+
+  if (env) {
+
+    INFO(PLUGIN_NAME " plugin: shutting down thread.");
+    if (stop_thread() < 0)
+      return -1;
+
+    mdb_env_close(env);
+  }
+
+  return 0;
+}
+
+static int presence_write(const data_set_t *ds, const value_list_t *vl,
+                          __attribute__((unused)) user_data_t *user_data) {
+
+  /* step aside quitely if not configured */
+  if (!env) {
+    return 0;
+  }
+
+  struct timeval tv_now = {0};
+
+  int rc = gettimeofday(&tv_now, /* struct timezone = */ NULL);
+  if (rc != 0) {
+    ERROR(PLUGIN_NAME " plugin: gettimeofday failed: %s", STRERRNO);
+    return -1;
+  }
+
+  /*
+   * Fast path - have we seen this host before?
+   *
+   * WHile this step is not strictly necessary, in the vast majority of
+   * cases we will have seen the host before, and we will have seen the
+   * host many times in the same second as each metric is written.
+   *
+   * The LMDB key value store offers very cheap lock free reads, allowing
+   * multiple threads to handle writes without any mutexes.
+   *
+   * This first step discards unnecessary writes as quickly as possible.
+   */
+
+  MDB_val key = {0}, data = {0};
+
+  key.mv_size = sizeof(vl->host);
+  key.mv_data = (void *)vl->host;
+
+  MDB_txn *txn = NULL;
+
+  rc = mdb_txn_begin(env, NULL, MDB_RDONLY, &txn);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_txn_begin returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  MDB_dbi dbi = {0};
+
+  rc = mdb_dbi_open(txn, NULL, 0, &dbi);
+  if (MDB_NOTFOUND == rc) {
+
+    /* brand new database, skip the get */
+
+  } else if (rc) {
+
+    /* error happened */
+
+    ERROR(PLUGIN_NAME " plugin: mdb_dbi_open returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    mdb_txn_abort(txn);
+    return -1;
+
+  } else {
+
+    /* existing database, do the get */
+
+    rc = mdb_get(txn, dbi, &key, &data);
+  }
+
+  int add = 0;
+
+  if (MDB_NOTFOUND == rc) {
```

Comment:
This code can be much more readable if you move up `add = 1` into L377 and
move block  `mdb_get failed` upper, below `mdb_get()`  call.

Then there will be no need in this second `if` construct.

Reply:
This suggestion breaks the first pass through this code.

There are three possible code paths:

- The very first time we see a host, we need to ignore the failure mdb_dbi_open(), and skip the get of mdb_get().
- All other times, we must mdb_dbi_open() and then mdb_get().
- An error has occurred.

In addition, code above "int add = 0;" is about open/get, while the code below is about whether we follow the fast path and skip the writing of the host.

These phases are kept separate from one another.